crates/jsonschema-referencing/src/anchors/mod.rs

```diff
@@ -1,33 +1,80 @@
-use std::sync::Arc;
+use std::{
+    hash::Hash,
+    sync::atomic::{AtomicPtr, Ordering},
+};
 
 use serde_json::Value;
 
 mod keys;
 
-use crate::{Draft, Error, Resolved, Resolver, Resource};
+use crate::{resource::ResourceRef, Draft, Error, Resolved, Resolver};
 pub(crate) use keys::{AnchorKey, AnchorKeyRef};
 
+#[derive(Debug)]
+pub(crate) struct AnchorName {
+    ptr: AtomicPtr<u8>,
+    len: usize,
+}
+
+impl AnchorName {
+    fn new(s: &str) -> Self {
+        Self {
+            ptr: AtomicPtr::new(s.as_ptr().cast_mut()),
+            len: s.len(),
+        }
+    }
+
+    fn as_str(&self) -> &str {
+        unsafe {
+            std::str::from_utf8_unchecked(std::slice::from_raw_parts(
+                self.ptr.load(Ordering::Relaxed),
+                self.len,
+            ))
+        }
+    }
+}
+
+impl Clone for AnchorName {
+    fn clone(&self) -> Self {
+        Self {
+            ptr: AtomicPtr::new(self.ptr.load(Ordering::Acquire)),
+            len: self.len,
+        }
+    }
+}
+
+impl Hash for AnchorName {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
```

Comment:
Given anchors are unique within the same schema, just pointer hashing could be fine